### DISCUSSION TITLE
How to define a one-to-many relationship from the parent side without foreign keys?

### DISCUSSION THREAD

**Discussion by HMaker** [sqlalchemy/sqlalchemy#11905] [Usage Questions] [answered]
I have a parent and child table where the child table can't have any foreign key to the parent (it's a [timescaledb limitation][1], the parent table is a "hypertable")

I hav tried the following
```py
import sqlalchemy as s
from sqlalchemy import orm

BaseSchema = orm.declarative_base()

class Parent1(BaseSchema):
    uuid = s.Column(s.UUID(True), nullable=False)
    # ...
    children = orm.relationship('Child', primaryjoin='Parent1.uuid == Child.parent_uuid')
    
class Parent2(BaseSchema):
    uuid = s.Column(s.UUID(True), nullable=False)
    # ...
    children = orm.relationship('Child', primaryjoin='Parent2.uuid == Child.parent_uuid')
    
class Child(BaseSchema):
    id = s.Column(s.Integer, primary_key=True, autoincrement=True)
    # ...
    parent_uuid = s.Column(s.UUID(True), nullable=False)
```

There are several possibles parents (they all concrete tables, inheritance is not possible). SQLAlchemy complains that it cannot find a foreign key to handle the primary join. I guess SA wishes the following:
```py
class Child(BaseSchema):
    id = s.Column(s.Integer, primary_key=True, autoincrement=True)
    # ...
    parent1_uuid = s.Column(s.UUID(True), nullable=False)
    parent = orm.relationship('Parent1', primaryjoin='Parent1.uuid == Child.parent1_uuid', foreign_keys='Child.parent1_uuid', back_populates='children')
 ```
 It requires me to add `parent1_uuid` and `parent2_uuid` columns to Child?

[1]: https://docs.timescale.com/use-timescale/latest/limitations/#hypertable-limitations

**Comment by CaselIT** [comment, marked as answer]
Hi,

Sqlalchemy is asking you to specify what is the foreign key. One of the option is to use `foreign()`. See example below:

```py
import sqlalchemy as s
from sqlalchemy import orm

BaseSchema = orm.declarative_base()

class Parent1(BaseSchema):
    __tablename__ = 'parent1'
    uuid = s.Column(s.UUID(True), nullable=False, primary_key=True)
    # ...
    children = orm.relationship('Child', primaryjoin='Parent1.uuid == foreign(Child.parent_uuid)')
    
class Parent2(BaseSchema):
    __tablename__ = 'parent2'
    uuid = s.Column(s.UUID(True), nullable=False, primary_key=True)
    # ...
    children = orm.relationship('Child', primaryjoin='Parent2.uuid == foreign(Child.parent_uuid)')
    
class Child(BaseSchema):
    __tablename__ = 'child'
    id = s.Column(s.Integer, primary_key=True, autoincrement=True)
    # ...
    parent_uuid = s.Column(s.UUID(True), nullable=False)


print(s.select(Parent1).join(Parent1.children))
print(s.select(Parent2).join(Parent2.children))
```

The code above will work correctly but it will raise the warning:
>  SAWarning: relationship 'Parent2.children' will copy column parent2.uuid to column child.parent_uuid, which conflicts with relationship(s): 'Parent1.children' (copies parent1.uuid to child.parent_uuid). If this is not the intention, consider if these relationships should be linked with back_populates, or if viewonly=True should be applied to one or more if they are read-only. For the less common case that foreign key constraints are partially overlapping, the orm.foreign() annotation can be used to isolate the columns that should be written towards.   To silence this warning, add the parameter 'overlaps="children"' to the 'Parent2.children' relationship. (Background on this warning at: https://sqlalche.me/e/21/qzyx) (This warning originated from the `configure_mappers()` process, which was invoked automatically in response to a user-initiated operation.)

That warning is correct since you have to relations that insist on the same column. You can either mark the relationship as read only or suppress the warning, just note that in this case care must be taken not to assign the same child to two parents

**Reply by HMaker** [reply]
Given the nature of UUIDs it's not possible (in theory) to a given child have 2 or more parents (Parent1 and Parent2 with the same UUID). Is that issue at the ORM level? SA requires foreign keys (Child.parent_uuid) to be unique across all parents' children in the Session?

I think supressing that warning with `overlaps="children"` is safe for my case.

**Reply by CaselIT** [reply]
No the issue the warning is telling you about is that you could place a children from parent1 into parent2.

**Reply by HMaker** [reply]
Ah ok, it warns about application bugs. Yes I got it, in fact if the same child instance is assigned to multiple parents, queries cascading to the children may overwrite each other.
